### PR TITLE
Downsample ILM action should skip non-time-series indices

### LINKED ISSUE #93123: Index is stuck in the `downsample` ILM action
### Elasticsearch Version

8.6.0

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Elasticsearch Service (ESS)

### Problem Description

Users may accidentally configure an ILM policy with the `downsample` [action](https://www.elastic.co/guide/en/elasticsearch/reference/current/ilm-downsample.html) where the ILM policy is attached to a regular index or a data stream (but not a [time series data stream (TSDS)](https://www.elastic.co/guide/en/elasticsearch/reference/current/tsds.html)).

In this situation, the regular index or backing index will remain stuck in the `downsample` ILM action (even after updating the ILM policy to remove the `downsample` action).

### Steps to Reproduce

1. Create a hot/warm deployment version 8.6.0 in ESS.

2. Create an ILM policy with a `downsample` action:
```
PUT _ilm/policy/my-lifecycle-policy
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover": {
            "max_docs": "2"
          },
          "downsample": {
            "fixed_interval": "1m"
          }
        }
      },
      "warm": {
        "min_age": "1m",
        "actions": {
          "allocate": {
            "number_of_replicas": 0
          }
        }
      },
      "delete": {
        "min_age": "2m",
        "actions": {
          "delete": {}
        }
      }
    }
  }
}
```

3. Update the `indices.lifecycle.poll_interval` for this current issue recreation:

```
PUT _cluster/settings
{
  "persistent": {
    "indices.lifecycle.poll_interval": "10s"
  }
}
```

4. Create a component template for mappings:
```
PUT _component_template/my-mappings
{
  "template": {
    "mappings": {
      "properties": {
        "@timestamp": {
          "type": "date",
          "format": "date_optional_time||epoch_millis"
        },
        "message": {
          "type": "wildcard"
        }
      }
    }
  },
  "_meta": {
    "description": "Mappings for @timestamp and message fields",
    "my-custom-meta-field": "More arbitrary metadata"
  }
}
```

5. Create a component template for index settings:
```
PUT _component_template/my-settings
{
  "template": {
    "settings": {
      "index.lifecycle.name": "my-lifecycle-policy"
    }
  },
  "_meta": {
    "description": "Settings for ILM",
    "my-custom-meta-field": "More arbitrary metadata"
  }
}
```

6. Create an index template:
```
PUT _index_template/my-index-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": { },
  "composed_of": [ "my-mappings", "my-settings" ],
  "priority": 500,
  "_meta": {
    "description": "Template for my time series data",
    "my-custom-meta-field": "More arbitrary metadata"
  }
}
```

7. Create the data stream:
```
PUT _data_stream/my-data-stream
```

8. Add some documents:
```
POST my-data-stream/_doc
{
  "@timestamp": "2023-01-22T00:00:01.000Z",
  "message": "document 1"
}

POST my-data-stream/_doc
{
  "@timestamp": "2023-01-22T00:00:02.000Z",
  "message": "document 2"
}

POST my-data-stream/_doc
{
  "@timestamp": "2023-01-22T00:00:03.000Z",
  "message": "document 3"
}
```

9. Verify that the documents are stored:
```
GET my-data-stream/_search
```

10. Verify that a new backing index has been created  (i.e `.ds-my-data-stream-2023.01.22-000002`) (will be created after the next poll interval)
```
GET _cat/indices?expand_wildcards=all
```

11. Check the ILM explain results of the first index (i.e `.ds-my-data-stream-2023.01.22-000001`) and observe that the index is stuck in the `downsample` action with the following exception:

```
Rollup requires setting [index.mode=time_series] for index [.ds-my-data-stream-2023.01.22-000001]
```

```
GET .ds-my-data-stream-2023.01.22-000001/_ilm/explain
```

```
{
  "indices": {
    ".ds-my-data-stream-2023.01.22-000001": {
      "index": ".ds-my-data-stream-2023.01.22-000001",
      "managed": true,
      "policy": "my-lifecycle-policy",
      "index_creation_date_millis": 1674363304040,
      "time_since_index_creation": "2.04m",
      "lifecycle_date_millis": 1674363319364,
      "age": "1.78m",
      "phase": "hot",
      "phase_time_millis": 1674363419410,
      "action": "downsample",
      "action_time_millis": 1674363319564,
      "step": "ERROR",
      "step_time_millis": 1674363419410,
      "failed_step": "rollup",
      "is_auto_retryable_error": true,
      "failed_step_retry_count": 9,
      "step_info": {
        "type": "exception",
        "reason": "Rollup requires setting [index.mode=time_series] for index [.ds-my-data-stream-2023.01.22-000001]"
      },
      "phase_execution": {
        "policy": "my-lifecycle-policy",
        "phase_definition": {
          "min_age": "0ms",
          "actions": {
            "downsample": {
              "fixed_interval": "1m"
            },
            "rollover": {
              "max_docs": 2
            }
          }
        },
        "version": 1,
        "modified_date_in_millis": 1674363301001
      }
    }
  }
}
```

12. Update the ILM policy to remove the `downsample` action:

```
PUT _ilm/policy/my-lifecycle-policy
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover": {
            "max_docs": "2"
          }
        }
      },
      "warm": {
        "min_age": "1m",
        "actions": {
          "allocate": {
            "number_of_replicas": 0
          }
        }
      },
      "delete": {
        "min_age": "2m",
        "actions": {
          "delete": {}
        }
      }
    }
  }
}
```

13. Check the ILM explain results of index `.ds-my-data-stream-2023.01.22-000001` and observe that the index remains stuck in the `downsample` action, despite the ILM policy update:

```
{
  "indices": {
    ".ds-my-data-stream-2023.01.22-000001": {
      "index": ".ds-my-data-stream-2023.01.22-000001",
      "managed": true,
      "policy": "my-lifecycle-policy",
      "index_creation_date_millis": 1674363304040,
      "time_since_index_creation": "2.79m",
      "lifecycle_date_millis": 1674363319364,
      "age": "2.54m",
      "phase": "hot",
      "phase_time_millis": 1674363439419,
      "action": "downsample",
      "action_time_millis": 1674363319564,
      "step": "ERROR",
      "step_time_millis": 1674363469445,
      "failed_step": "rollup",
      "is_auto_retryable_error": true,
      "failed_step_retry_count": 14,
      "step_info": {
        "type": "exception",
        "reason": "Rollup requires setting [index.mode=time_series] for index [.ds-my-data-stream-2023.01.22-000001]"
      },
      "phase_execution": {
        "policy": "my-lifecycle-policy",
        "phase_definition": {
          "min_age": "0ms",
          "actions": {
            "downsample": {
              "fixed_interval": "1m"
            },
            "rollover": {
              "max_docs": 2
            }
          }
        },
        "version": 1,
        "modified_date_in_millis": 1674363301001
      }
    }
  }
}
```

### Logs (if relevant)

_No response_

### Workaround

I am not aware of any workarounds for now. To be discussed with @elastic/es-data-management .

### PR DESCRIPTION
**Pull request by csoulios**
Make the `downsample` action skip the non-time-series indices

As discussed in https://github.com/elastic/elasticsearch/issues/93123#issuecomment-1409919490:

> ...  we think the downsample action should skip backing indices that are not part of a time series data stream. We do a similar approach with the unfollow action where non follower indices are skipped. And this allows to have the same policy for both regular data streams and tsdb data streams.

Closes https://github.com/elastic/elasticsearch/issues/93123

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/UpdateRollupIndexPolicyStep.java [hunk 1]
```diff
@@ -1,92 +0,0 @@
...
-import org.elasticsearch.common.Strings;
-import org.elasticsearch.common.settings.Settings;
-import org.elasticsearch.core.TimeValue;
-
-import java.util.Objects;
-
-/**
- * Updates the lifecycle policy for the rollup index for the original/currently managed index
- */
-public class UpdateRollupIndexPolicyStep extends AsyncActionStep {
```
[the remark this answers is not in the document]
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/UpdateRollupIndexPolicyStep.java:26]: Ah nice, thanks for deleting this

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/DownsampleActionIT.java [hunk 2]
```diff
@@ -109,16 +109,24 @@ public void refreshAbstractions() {
         );
     }
 
-    private void createIndex(String index, String alias) throws IOException {
+    @Before
+    public void updateClusterSettings() throws IOException {
+        updateClusterSettings(client(), Settings.builder().put("indices.lifecycle.poll_interval", "5s").build());
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/DownsampleActionIT.java:114]: Is this something we want? What's the purpose of delaying the poll interval to later than the 1 second we configure it in tests?

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/DownsampleActionIT.java [hunk 3]
```diff
@@ -280,6 +287,20 @@ public void testTsdbDataStreams() throws Exception {
...
+        index(client(), index, true, null, "@timestamp", "2020-01-01T05:10:00Z", "volume", 11.0, "metricset", randomAlphaOfLength(5));
+
+        String phaseName = randomFrom("warm", "cold");
+        createNewSingletonPolicy(client(), policy, phaseName, new DownsampleAction(ConfigTestHelpers.randomInterval()));
+        updatePolicy(client(), index, policy);
+
+        String rollupIndex = waitAndGetRollupIndexName(client(), index);
+        assertNull("Rollup index should not have been created", rollupIndex);
+        assertBusy(() -> assertTrue("Source index should not have been deleted", indexExists(index)));
+        assertBusy(() -> assertThat(getStepKeyForIndex(client(), index), equalTo(PhaseCompleteStep.finalStep(phaseName).getKey())));
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/DownsampleActionIT.java:301]: These assertions make sense, but they are written in a way that would make the test idle for 60 seconds until `waitAndGetRollupIndexName` timeouts.

Would it make sense to have the final step assertion be the first one ?
i.e. 
```
assertBusy(() -> assertThat(getStepKeyForIndex(client(), index), equalTo(PhaseCompleteStep.finalStep(phaseName).getKey())));
```
Followed by regular assertions (they don't need to be `assertBusy` anymore as ILM is in the final step already)
e.g.
```
assertNull("Rollup index should not have been created", rollupIndex);
assertTrue("Source index should not have been deleted", indexExists(index));
```

**Review by andreidan — COMMENTED** [review]
Thanks for working on this Christos

Left a few minor comments

File: x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/DownsampleActionIT.java (hunk 2 above)
Comment (csoulios) [reply x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/DownsampleActionIT.java:114]: For some reason the `@After cleanUpCluster()` method sets the following: 
```
[2023-03-29T13:19:01,928][INFO ][o.e.c.s.ClusterSettings  ] [runTask-0] updating [indices.lifecycle.poll_interval] from [5s] to [10m]
``` 

So, I need to reset it at my `@Before` test

Comment (andreidan) [reply x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/DownsampleActionIT.java:114]: I just ran the test on your branch (after having removed this @Before method) and I don't see the setting being changed to `10m`.

Gradle configures the poll interval to `1000ms` for these test clusters https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ilm/qa/multi-node/build.gradle#L29 (`elasticsearch.yml` for all nodes in the cluster will contain this setting)

This is my diff for your branch 
```
--- a/x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/DownsampleActionIT.java
+++ b/x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/DownsampleActionIT.java
@@ -109,11 +109,6 @@ public class DownsampleActionIT extends ESRestTestCase {
         );
     }

-    @Before
-    public void updateClusterSettings() throws IOException {
-        updateClusterSettings(client(), Settings.builder().put("indices.lifecycle.poll_interval", "5s").build());
-    }
-
     private void createIndex(String index, String alias, boolean isTimeSeries) throws IOException {
         Settings.Builder settings = Settings.builder()
             .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
```

Could you remove this and try again please? I'm fairly convinced we don't need it

**Review by martijnvg — APPROVED** [review]
LGTM

**Pull request merged**